src/gather_classes.rs

```diff
@@ -0,0 +1,417 @@
+    for (block_num, path) in file_paths.into_iter() {
+        info!(
+            "{}\t{}\t{}",
+            get_stat("Processed", processed),
+            get_stat("Successes", processed - failures),
+            get_stat("Failures", failures)
+        );
+        info!(
+            "Processing path: `{}`",
+            path.file_name().unwrap().to_str().unwrap()
+        );
+
+        let maybe_block_report = match force_generate {
+            false => {
+                try_deserialize(&path_for_report(block_num)).or_else(|_| generate_report(&path))
+            }
+            true => generate_report(&path),
+        };
+
+        match maybe_block_report {
+            Err(err) => {
+                warn!("Failed to process file with err: {err}");
+                failures += 1;
+            }
+            Ok(block_report) => {
+                let _ = try_serialize_block_report(block_num, &block_report, force_generate)
+                    .map_err(|err| warn!("{err:?}"));
+                processed += 1;
+
+                let update_result = overall_report.update(&block_report);
+                match update_result {
+                    Err(err) => {
+                        info!("{err}");
+                        failures += 1;
+                    }
+                    Ok(updates) => {
+                        info!("Updates to overall report from block {block_num}: {updates}");
+                        let _ = try_serialize(path_for_overall_report(), &overall_report)
+                            .inspect_err(|err| warn!("{err}"));
+                    }
+                }
+            }
+        }
+    }
+    info!(
+        "{}\t{}\t{}",
+        get_stat("Processed", processed),
+        get_stat("Successes", processed - failures),
+        get_stat("Failures", failures)
+    );
+}
+
+fn generate_report<P>(path: P) -> Result<ClassHashesReport, anyhow::Error>
+where
+    P: AsRef<Path>,
+{
+    let path_str = path.as_ref().to_str().expect("failed to unwrap path");
+
+    let mut comparison_file = OpenOptions::new()
+        .read(true)
+        .open(path.as_ref())
+        .context(format!("Reading comparison file: '{}'", path_str,))
+        .inspect_err(|err| {
+            warn!("Error reading file: {err}");
+        })?;
+
+    let mut data = String::new();
+    let _data_len = comparison_file.read_to_string(&mut data)?;
+    let json_obj: Value = serde_json::from_str(data.as_str()).expect("unwrapping json_obj");
+
+    let json_block_num = block_num_from_json(&json_obj);
+    let path_block_num = block_num_from_path(path_str);
+    let block_num = val_or_err(path_block_num, json_block_num)?;
+
+    let mut report = ClassHashesReport::new();
+    update_report(&mut report, &json_obj, block_num)?;
+    assert!(report.blocks.len() == 1 && report.blocks.contains(&block_num));
+    Ok(report)
```

Comment:
I understand that perhaps the initial idea was to merge different Class Hash Report but I don't think this adapts well to the current problem. In this problem, we are having one `ClassHashesReport` and several `ComparisonBlocks`. 

I think is better rather creating `ClassHashesReport` per block and then merging it, it makes more sense to simply add the blocks to an already existing `ClassHashesReport` without needing to create a new one each time.

Reply:
Removed `generate_report<P>` and modified `process_compare_paths` accordingly.